[CIR][X86] Implement lowering for pmuldq / pmuludq builtins

GeneraluseAI opened the pull request:
part of [#167765](https://github.com/llvm/llvm-project/issues/167765)

This patch adds CIR codegen support for X86 pmuldq and pmuludq operations, covering the signed and unsigned variants across all supported vector widths. The builtins now lower to the expected CIR representation matching the semantics of the corresponding LLVM intrinsics.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltinX86.cpp:
```diff
@@ -115,6 +115,47 @@ static mlir::Value emitX86MaskLogic(CIRGenBuilderTy &builder,
+        cir::VecSplatOp::create(builder, loc, ty, shiftAmt.getResult());
+    mlir::Value shiftSplatValue = shiftSplatVecOp.getResult();
+    lhs = builder.createBitcast(loc, ops[0], ty);
+    rhs = builder.createBitcast(loc, ops[1], ty);
```
These also should be hoisted out of the `if-else`.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltinX86.cpp:
```diff
@@ -115,6 +115,47 @@ static mlir::Value emitX86MaskLogic(CIRGenBuilderTy &builder,
+  mlir::Type ty = ops[0].getType();
+  unsigned tyPrimitiveSizeInBits =
+      cgf.cgm.getDataLayout().getTypeSizeInBits(ty);
+  mlir::Value lhs, rhs;
```
```suggestion
  mlir::Value lhs = builder.createBitcast(loc, ops[0], ty);
  mlir::Value rhs = builder.createBitcast(loc, ops[1], ty);
```

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltinX86.cpp:
```diff
@@ -851,12 +892,18 @@ mlir::Value CIRGenFunction::emitX86BuiltinExpr(unsigned builtinID,
   case X86::BI__builtin_ia32_pmuldq128:
   case X86::BI__builtin_ia32_pmuldq256:
   case X86::BI__builtin_ia32_pmuldq512:
+    return emitX86Muldq(*this, expr, /*IsSigned*/ true, ops);
```
```suggestion
    return emitX86Muldq(builder, getLoc(expr->getExprLoc), /*isSigned=*/true, ops);
```

GeneraluseAI commented on clang/lib/CIR/CodeGen/CIRGenBuiltinX86.cpp:
```diff
@@ -115,6 +115,47 @@ static mlir::Value emitX86MaskLogic(CIRGenBuilderTy &builder,
                                ops[0].getType());
 }
 
+static mlir::Value emitX86Muldq(CIRGenFunction &cgf, const CallExpr *expr,
```
@andykaylor  Hi I just modified to compute tyPrimitiveSizeInBits before calling emitX86Muldq and pass it in as an extra parameter. Don't know If this is ok, please let me know if you have a better solution or you this is acceptable. Thanks!

andykaylor reviewed (approved):
lgtm

GeneraluseAI committed:
[CIR][X86] Implement lowering for pmuldq / pmuludq builtins

This patch adds CIR codegen support for X86 pmuldq and pmuludq operations,
covering the signed and unsigned variants across all supported vector widths.
The builtins now lower to the expected CIR representation matching the
semantics of the corresponding LLVM intrinsics.
